File: swift-bridge/Sources/SpriteKitBridge/Constraint.swift
import SpriteKit

@_cdecl("sk_range_new")
public func sk_range_new(_ lower: Double, _ upper: Double) -> UnsafeMutableRawPointer? {
    skRetain(SKRange(lowerLimit: CGFloat(lower), upperLimit: CGFloat(upper)))
}

@_cdecl("sk_range_with_lower_limit")
public func sk_range_with_lower_limit(_ lower: Double) -> UnsafeMutableRawPointer? {
    skRetain(SKRange(lowerLimit: CGFloat(lower)))
}

@_cdecl("sk_range_with_upper_limit")
public func sk_range_with_upper_limit(_ upper: Double) -> UnsafeMutableRawPointer? {
    skRetain(SKRange(upperLimit: CGFloat(upper)))
}

@_cdecl("sk_range_with_constant")
public func sk_range_with_constant(_ value: Double) -> UnsafeMutableRawPointer? {
    skRetain(SKRange(constantValue: CGFloat(value)))
}

@_cdecl("sk_range_with_variance")
public func sk_range_with_variance(_ value: Double, _ variance: Double) -> UnsafeMutableRawPointer? {
    skRetain(SKRange(value: CGFloat(value), variance: CGFloat(variance)))
}

@_cdecl("sk_range_with_no_limits")
public func sk_range_with_no_limits() -> UnsafeMutableRawPointer? {
    skRetain(SKRange())
}

@_cdecl("sk_range_get_lower_limit")
public func sk_range_get_lower_limit(_ rangeHandle: UnsafeMutableRawPointer?) -> Double {
    guard let range: SKRange = skBorrow(rangeHandle) else { return 0 }
    return Double(range.lowerLimit)
}

@_cdecl("sk_range_set_lower_limit")
public func sk_range_set_lower_limit(_ rangeHandle: UnsafeMutableRawPointer?, _ lower: Double) {
    guard let range: SKRange = skBorrow(rangeHandle) else { return }
    range.lowerLimit = CGFloat(lower)
}

@_cdecl("sk_range_get_upper_limit")
public func sk_range_get_upper_limit(_ rangeHandle: UnsafeMutableRawPointer?) -> Double {
    guard let range: SKRange = skBorrow(rangeHandle) else { return 0 }
    return Double(range.upperLimit)
}

@_cdecl("sk_range_set_upper_limit")
public func sk_range_set_upper_limit(_ rangeHandle: UnsafeMutableRawPointer?, _ upper: Double) {
    guard let range: SKRange = skBorrow(rangeHandle) else { return }
    range.upperLimit = CGFloat(upper)
}

@_cdecl("sk_constraint_position_x")
public func sk_constraint_position_x(_ rangeHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let range: SKRange = skBorrow(rangeHandle) else { return nil }
    return skRetain(SKConstraint.positionX(range))
}

@_cdecl("sk_constraint_position_y")
public func sk_constraint_position_y(_ rangeHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let range: SKRange = skBorrow(rangeHandle) else { return nil }
    return skRetain(SKConstraint.positionY(range))
}

@_cdecl("sk_constraint_position_xy")
public func sk_constraint_position_xy(_ xRangeHandle: UnsafeMutableRawPointer?, _ yRangeHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let xRange: SKRange = skBorrow(xRangeHandle),
          let yRange: SKRange = skBorrow(yRangeHandle)
    else { return nil }
    return skRetain(SKConstraint.positionX(xRange, y: yRange))
}

@_cdecl("sk_constraint_distance_to_node")
public func sk_constraint_distance_to_node(_ rangeHandle: UnsafeMutableRawPointer?, _ nodeHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let range: SKRange = skBorrow(rangeHandle),
          let node: SKNode = skBorrow(nodeHandle)
    else { return nil }
    return skRetain(SKConstraint.distance(range, to: node))
}

@_cdecl("sk_constraint_distance_to_point")
public func sk_constraint_distance_to_point(_ rangeHandle: UnsafeMutableRawPointer?, _ x: Double, _ y: Double) -> UnsafeMutableRawPointer? {
    guard let range: SKRange = skBorrow(rangeHandle) else { return nil }
    return skRetain(SKConstraint.distance(range, to: CGPoint(x: x, y: y)))
}

@_cdecl("sk_constraint_distance_to_point_in_node")
public func sk_constraint_distance_to_point_in_node(_ rangeHandle: UnsafeMutableRawPointer?, _ x: Double, _ y: Double, _ nodeHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let range: SKRange = skBorrow(rangeHandle),
          let node: SKNode = skBorrow(nodeHandle)
    else { return nil }
    return skRetain(SKConstraint.distance(range, to: CGPoint(x: x, y: y), in: node))
}

@_cdecl("sk_constraint_z_rotation")
public func sk_constraint_z_rotation(_ rangeHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let range: SKRange = skBorrow(rangeHandle) else { return nil }
    return skRetain(SKConstraint.zRotation(range))
}

@_cdecl("sk_constraint_orient_to_node")
public func sk_constraint_orient_to_node(_ nodeHandle: UnsafeMutableRawPointer?, _ offsetHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let node: SKNode = skBorrow(nodeHandle),
          let offset: SKRange = skBorrow(offsetHandle)
    else { return nil }
    return skRetain(SKConstraint.orient(to: node, offset: offset))
}

@_cdecl("sk_constraint_orient_to_point")
public func sk_constraint_orient_to_point(_ x: Double, _ y: Double, _ offsetHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let offset: SKRange = skBorrow(offsetHandle) else { return nil }
    return skRetain(SKConstraint.orient(to: CGPoint(x: x, y: y), offset: offset))
}

@_cdecl("sk_constraint_orient_to_point_in_node")
public func sk_constraint_orient_to_point_in_node(_ x: Double, _ y: Double, _ nodeHandle: UnsafeMutableRawPointer?, _ offsetHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let node: SKNode = skBorrow(nodeHandle),
          let offset: SKRange = skBorrow(offsetHandle)
    else { return nil }
    return skRetain(SKConstraint.orient(to: CGPoint(x: x, y: y), in: node, offset: offset))
}

@_cdecl("sk_constraint_get_enabled")
public func sk_constraint_get_enabled(_ constraintHandle: UnsafeMutableRawPointer?) -> Bool {
    guard let constraint: SKConstraint = skBorrow(constraintHandle) else { return false }
    return constraint.enabled
}

@_cdecl("sk_constraint_set_enabled")
public func sk_constraint_set_enabled(_ constraintHandle: UnsafeMutableRawPointer?, _ enabled: Bool) {
    guard let constraint: SKConstraint = skBorrow(constraintHandle) else { return }
    constraint.enabled = enabled
}

@_cdecl("sk_constraint_get_reference_node")
public func sk_constraint_get_reference_node(_ constraintHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let constraint: SKConstraint = skBorrow(constraintHandle), let referenceNode = constraint.referenceNode else { return nil }
    return skRetain(referenceNode)
}

@_cdecl("sk_constraint_set_reference_node")
public func sk_constraint_set_reference_node(_ constraintHandle: UnsafeMutableRawPointer?, _ nodeHandle: UnsafeMutableRawPointer?) {
    guard let constraint: SKConstraint = skBorrow(constraintHandle) else { return }
    let node: SKNode? = skBorrow(nodeHandle)
    constraint.referenceNode = node
}

File: src/constraint.rs
use apple_cf::cg::CGPoint;

use crate::ffi;
use crate::node::{AsNode, Node};
use crate::private::handle_type;

handle_type!(ConstraintRange);
handle_type!(Constraint);

impl ConstraintRange {
    #[must_use]
    pub fn with_limits(lower: f64, upper: f64) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_range_new(lower, upper)) }
    }

    #[must_use]
    pub fn with_lower_limit(lower: f64) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_range_with_lower_limit(lower)) }
    }

    #[must_use]
    pub fn with_upper_limit(upper: f64) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_range_with_upper_limit(upper)) }
    }

    #[must_use]
    pub fn with_constant(value: f64) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_range_with_constant(value)) }
    }

    #[must_use]
    pub fn with_variance(value: f64, variance: f64) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_range_with_variance(value, variance)) }
    }

    #[must_use]
    pub fn no_limits() -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_range_with_no_limits()) }
    }

    #[must_use]
    pub fn lower_limit(&self) -> f64 {
        unsafe { ffi::sk_range_get_lower_limit(self.ptr) }
    }

    pub fn set_lower_limit(&self, lower: f64) {
        unsafe { ffi::sk_range_set_lower_limit(self.ptr, lower) };
    }

    #[must_use]
    pub fn upper_limit(&self) -> f64 {
        unsafe { ffi::sk_range_get_upper_limit(self.ptr) }
    }

    pub fn set_upper_limit(&self, upper: f64) {
        unsafe { ffi::sk_range_set_upper_limit(self.ptr, upper) };
    }
}

impl Constraint {
    #[must_use]
    pub fn position_x(range: &ConstraintRange) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_constraint_position_x(range.as_ptr())) }
    }

    #[must_use]
    pub fn position_y(range: &ConstraintRange) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_constraint_position_y(range.as_ptr())) }
    }

    #[must_use]
    pub fn position_xy(x_range: &ConstraintRange, y_range: &ConstraintRange) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_constraint_position_xy(x_range.as_ptr(), y_range.as_ptr())) }
    }

    #[must_use]
    pub fn distance_to_node<N: AsNode>(range: &ConstraintRange, node: &N) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_constraint_distance_to_node(range.as_ptr(), node.as_node_ptr())) }
    }

    #[must_use]
    pub fn distance_to_point(range: &ConstraintRange, point: CGPoint) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_constraint_distance_to_point(range.as_ptr(), point.x, point.y)) }
    }

    #[must_use]
    pub fn distance_to_point_in_node<N: AsNode>(
        range: &ConstraintRange,
        point: CGPoint,
        node: &N,
    ) -> Option<Self> {
        unsafe {
            Self::from_raw(ffi::sk_constraint_distance_to_point_in_node(
                range.as_ptr(),
                point.x,
                point.y,
                node.as_node_ptr(),
            ))
        }
    }

    #[must_use]
    pub fn z_rotation(range: &ConstraintRange) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_constraint_z_rotation(range.as_ptr())) }
    }

    #[must_use]
    pub fn orient_to_node<N: AsNode>(node: &N, offset: &ConstraintRange) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_constraint_orient_to_node(node.as_node_ptr(), offset.as_ptr())) }
    }

    #[must_use]
    pub fn orient_to_point(point: CGPoint, offset: &ConstraintRange) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_constraint_orient_to_point(point.x, point.y, offset.as_ptr())) }
    }

    #[must_use]
    pub fn orient_to_point_in_node<N: AsNode>(
        point: CGPoint,
        node: &N,
        offset: &ConstraintRange,
    ) -> Option<Self> {
        unsafe {
            Self::from_raw(ffi::sk_constraint_orient_to_point_in_node(
                point.x,
                point.y,
                node.as_node_ptr(),
                offset.as_ptr(),
            ))
        }
    }

    #[must_use]
    pub fn is_enabled(&self) -> bool {
        unsafe { ffi::sk_constraint_get_enabled(self.ptr) }
    }

    pub fn set_enabled(&self, enabled: bool) {
        unsafe { ffi::sk_constraint_set_enabled(self.ptr, enabled) };
    }

    #[must_use]
    pub fn reference_node(&self) -> Option<Node> {
        unsafe { Node::from_raw(ffi::sk_constraint_get_reference_node(self.ptr)) }
    }

    pub fn set_reference_node<N: AsNode>(&self, node: Option<&N>) {
        unsafe {
            ffi::sk_constraint_set_reference_node(
                self.ptr,
                node.map_or(core::ptr::null_mut(), AsNode::as_node_ptr),
            );
        };
    }
}
